v2/inc/c_logging/log_context.h

```diff
@@ -0,0 +1,105 @@
+#define LOG_CONTEXT_H
+
+#ifdef __cplusplus
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#else
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#endif
+
+#include "macro_utils/macro_utils.h"
+
+// This code is intentionally a spike only to showcase an implementation
+// Code with tests shall be written after comments on the general idea are received
+
+typedef struct LOG_CONTEXT_TAG
+{
+    char* context_string;
+} LOG_CONTEXT;
+
+typedef struct LOG_CONTEXT_TAG* LOG_CONTEXT_HANDLE;
+
+#define EXPAND_LOG_CONTEXT_FIELD(field_name, field_PRI, field_values) \
+    chars_written = sprintf(pos, "%s=" field_PRI " ", field_name, field_values); \
+    pos += (size_t)chars_written; \
+
+#define STRINGIFY_LOG_CONTEXT_FIELD(field_desc) \
+    MU_C2(EXPAND_, field_desc)
+
+// macro that can be used to create a context on the stack
+#define LOCAL_LOG_CONTEXT_DEFINE(log_context, parent_context, ...) \
+    char MU_C2(log_string_, log_context)[LOG_MAX_MESSAGE_LENGTH]; \
+    uint32_t MU_C2(parent_context_string_length_, log_context) = (uint32_t)log_context_get_string_length(parent_context); \
+    { \
+        const char* parent_context_string = log_context_get_string(parent_context); \
+        if (parent_context_string != NULL) \
+        { \
+            (void)memcpy(MU_C2(log_string_, log_context), parent_context_string, MU_C2(parent_context_string_length_, log_context)); \
+        } \
+        char* pos = MU_C2(log_string_, log_context) + MU_C2(parent_context_string_length_, log_context); \
+        int chars_written = 0; \
+        MU_FOR_EACH_1(STRINGIFY_LOG_CONTEXT_FIELD, __VA_ARGS__) \
+    } \
+    LOG_CONTEXT log_context = { MU_C2(log_string_, log_context) };
+
+#define EXPAND_GET_LENGTH_LOG_CONTEXT_FIELD(field_name, field_PRI, field_values) \
+    chars_written += snprintf(NULL, 0, "%s=" field_PRI " ", field_name, field_values); \
+
+#define GET_LOG_CONTEXT_FIELD_LENGTH(field_desc) \
+    MU_C2(EXPAND_GET_LENGTH_, field_desc)
+
+static size_t log_context_get_string_length(LOG_CONTEXT_HANDLE log_context)
+{
+    return (log_context == NULL) ? 0 : strlen(log_context->context_string);
+}
+
+static const char* log_context_get_string(LOG_CONTEXT_HANDLE log_context)
+{
+    return (log_context == NULL) ? NULL : log_context->context_string;
+}
+
+// macro that can be used to create a dynamically allocated context
+#define LOG_CONTEXT_CREATE(dest_log_context, parent_context, ...) \
+    uint32_t MU_C2(parent_context_string_length_, dest_log_context) = (uint32_t)log_context_get_string_length(parent_context); \
+    int MU_C2(chars_to_write_, dest_log_context); \
+    { \
+        int chars_written = 0; \
+        MU_FOR_EACH_1(GET_LOG_CONTEXT_FIELD_LENGTH, __VA_ARGS__) \
+        MU_C2(chars_to_write_, dest_log_context) = chars_written + MU_C2(parent_context_string_length_, dest_log_context) + 1; \
+    } \
+    dest_log_context = malloc(sizeof(LOG_CONTEXT) + MU_C2(chars_to_write_, dest_log_context) + 1); \
```

Comment:
>malloc [](http://example.com/codeflow?start=23&length=6)

malloc_2? or it doesn't exist in this context/\?

Reply:
We don't have it in this context.